k2/csrc/top_sort.cu

```diff
@@ -337,6 +363,47 @@ class TopSorter {
 
     std::vector<std::unique_ptr<Ragged<int32_t>>> iters;
     iters.push_back(GetInitialBatch());
+    // the start state in each Fsa must be present in the first batch
```

Comment:
If this extra code is purely checking code it should be clearly marked as such, so that a person trying to understand this function does not waste time trying to understand it.
I would recommend that you (preferably) move it to a separate function, or at least put it in a code block like as follows:
```
  {  // This block just checks that all non-empty FSAs in the input have their start state in their
     // first batch.
  }
```
However, I think it would be easy to check this in a single kernel, probably with num_fsas as its 'n' parameter
(you could locate the 1st state in the batch for each FSA and see if it's the start state).
You could just do a K2_CHECK_EQ inside the kernel if you want to make it easy to write.

Reply:
Oh, yes, as our implementation of `GetInitBatch` will always keep the start state (if its indegree is zero) as the 1st state. Will update, thanks!

Reply:
Updated as you said, OK to merge if there's no other issues, will go to fix the bug in the general `IsRandEquivalent` you metioned